Fix astropy so that it can be used in app bundles, and add CI to prevent regression

astrofrog opened the pull request:
With applications like [pyinstaller](https://www.pyinstaller.org/), it's easy to build application bundles for different platforms. Unfortunately, astropy doesn't work out of the box with these and requires some tweaking due to various issues. In the past, @embray actually opened a PR to fix this (https://github.com/astropy/astropy/pull/960) but this was never finished.

Since it would be nice for this to work cleanly, I've decided to start off by adding a CircleCI job that will test that this works correctly and that we don't regress in future. I'll then work on fixing issues, and will see if I can cherry pick or adapt any of the solutions from https://github.com/astropy/astropy/pull/960. So this is just a WIP for now.

UPDATE: this is now ready, see https://github.com/astropy/astropy/pull/8795#issuecomment-623552977

EDIT: Fix #7052

astrofrog commented:
Some of the fixes will much easier to do by adding a ``hook-astropy.py`` into PyInstaller, so I'll work on that.

bsipocz commented:
@astrofrog - I'm bumping the milestone to the next bugfix, to help us see how we stand. Of course if any from there is ready before the release, they can be moved back to 4.0.

astrofrog commented:
Sounds good, I definitely won't have time to finish this for 4.0.

bsipocz commented:
is this still something on the table post APE17?

astrofrog commented:
Yes this would still be nice to get working, I just haven't had time

bsipocz commented:
no worries, I just remilestoned it then. We can always move it back to the bugfix milestone when it's merged.

mhvk commented:
Not release-critical, so removing milestone.

astrofrog commented:
This is now ready for review - all required fixes to make astropy work properly in bundles were done in separate PRs over the last year or so, this now just add a tox environment and associated Travis cron job to make sure that we don't break this capability. The Travis job is 7 minutes in total, so not a huge overhead, and I've only enabled it as a cron job to avoid slowing down any PRs. I had a temporary commit here to make sure it ran correctly, but have now removed the commit so the job won't show up here (and CI should therefore pass anyway).

A side bonus is that this makes sure that the tests are able to be extracted as-is and run against an installed version of astropy (this was needed since pytest cannot discover tests in application bundles) - normally this should work, but there were a few relative imports that have crept in since we switched to absolute imports. Relative imports to the same directory are fine, but this fixes ones where the relative import went one or more levels up (which shouldn't be there).

In any case, this is now ready for review and can be backported to 4.0.x.

astrofrog commented:
I think with this merged we can close https://github.com/astropy/astropy/issues/7052

mhvk commented on .travis.yml:
```diff
@@ -188,6 +188,14 @@ matrix:
           stage: Cron tests
           env: SETUP_METHOD='apt'
 
+        # Regularly make sure that astropy can be used
```
Comment incomplete - "in an Application bundle"?

astrofrog replied:
Oops thanks :)

astrofrog commented:
The success over in #7052 is due to fixes that were done in other PRs in the process of doing this one - this current PR essentially just adds the regression test :smile:

astrofrog commented:
Since this has been approved, I'll go ahead and merge so that it doesn't go stale :smile:

embray commented:
Wow, my first attempt to solve this was 7 years ago now.  Amazing to see this finally done :)

astrofrog commented:
@embray thank you for starting this!
